tests/test_vrf.py

```diff
@@ -0,0 +1,1395 @@
+            # Verify bgp sessions are established
+            assert bgp_info['ipv4Unicast']['peers'][str(ptf_speaker_ip.ip)]['state'] == 'Established', "Bgp peer {} should be Established!".format(ptf_speaker_ip.ip)
+            # Verify accepted prefixes of the dynamic neighbors are correct
+            assert bgp_info['ipv4Unicast']['peers'][str(ptf_speaker_ip.ip)]['prefixReceivedCount'] == 1
+
+
+class TestVrfWarmReboot():
+    @pytest.fixture(scope="class", autouse=True)
+    def setup_vrf_warm_reboot(self, ptfhost, testbed):
+        # -------- Setup ----------
+        gen_vrf_fib_file('Vrf1', testbed, ptfhost,
+                    dst_intfs=['PortChannel0001', 'PortChannel0002'],
+                    dst_file='/tmp/vrf1_fib.txt',
+                    limited_podset_number=50,
+                    limited_tor_number=16
+                    )
+
+        # -------- Testing ----------
+        yield
+
+        # -------- Teardown ----------
+        pass
+
+    def test_vrf_swss_warm_reboot(self, duthost, ptfhost, host_facts, testbed, cfg_facts):
+        # enable swss warm-reboot
+        duthost.shell("config warm_restart enable swss")
+
+        exc_que = Queue.Queue()
+        params = {
+                    'ptf_runner': ptf_runner,
+                    'exc_queue': exc_que,  # use for store exception infos
+                    'host': ptfhost,
+                    'testdir': 'ptftests',
+                    'testname': 'vrf_test.FibTest',
+                    'platform_dir': 'ptftests',
+                    'params': {
+                                'testbed_type': testbed['topo'],
+                                'router_mac': host_facts['ansible_Ethernet0']['macaddress'],
+                                'fib_info': "/tmp/vrf1_fib.txt",
+                                'src_ports': g_vars['vrf_member_port_indices']['Vrf1']
+                              },
+                    'log_file': "/tmp/vrf_swss_warm_test.FibTest.log"
+                 }
+
+        traffic_in_bg = threading.Thread(target=ex_ptf_runner, kwargs=params)
+
+        # send background traffic
+        traffic_in_bg.start()
+
+        # start swss warm-reboot
+        duthost.shell("service swss restart")
+
+        # wait until background traffic finished
+        traffic_in_bg.join()
+
+        passed = True
+        if exc_que.qsize() != 0:
+            passed = False
+            exc_type, exc_obj, exc_trace = exc_que.get()
+        assert passed == True, "Traffic Test Failed \n {}".format(str(exc_obj))
+
+        # wait until components finish reconcile
+        tbd_comp_list = finalize_warmboot(duthost)
+        assert len(tbd_comp_list) == 0, "Some components didn't finish reconcile: {} ...".format(tbd_comp_list)
+
+        # basic check after warm reboot
+        duthost.shell("docker exec -i syncd ps aux | grep /usr/bin/syncd")
+        duthost.shell("docker exec -i swss ps aux | grep orchagent")
```

Comment:
maybe you can reuse critical service check function: https://github.com/Azure/sonic-mgmt/blob/master/tests/common/devices.py#L154

Reply:
Use `wait_until` + `critical_service_check` for better handling this.